dotCMS/src/main/java/com/dotcms/publisher/util/dependencies/DependencyManager.java

```diff
@@ -0,0 +1,417 @@
+									+ " does NOT have working or live version, not Pushed");
+				} else {
+					add(scheme, PusheableAsset.WORKFLOW);
+				}
+			} else if (asset.getType().equals(PusheableAsset.LANGUAGE.getType())) {
+				Language language = APILocator.getLanguageAPI()
+						.getLanguage(asset.getAsset());
+				if (language == null || !UtilMethods.isSet(language.getLanguage())) {
+					Logger.warn(getClass(), "Language id: "
+							+ (asset.getAsset() != null ? asset.getAsset()
+							: "N/A")
+							+ " is not present in the database, not Pushed");
+				} else {
+					add(language, PusheableAsset.LANGUAGE);
+				}
+			} else if (asset.getType().equals(PusheableAsset.RULE.getType())) {
+				Rule rule = APILocator.getRulesAPI()
+						.getRuleById(asset.getAsset(), user, false);
+				if (rule != null && StringUtils.isNotBlank(rule.getId())) {
+					add(rule, PusheableAsset.RULE);
+				} else {
+					Logger.warn(getClass(), "Rule id: "
+							+ (asset.getAsset() != null ? asset.getAsset()
+							: "N/A")
+							+ " is not present in the database, not Pushed.");
+				}
+			}
+		}
+
+		if (UtilMethods.isSet(config.getLuceneQueries())) {
+			List<String> contentIds = PublisherUtil.getContentIds(config.getLuceneQueries());
+			contentIds.removeIf(c -> publisherFilter.doesExcludeQueryContainsContentletId(c));
+
+			for (String id : contentIds) {
+				final Identifier ident = APILocator.getIdentifierAPI().find(id);
+				final List<Contentlet> contentlets = APILocator.getContentletAPI()
+						.findAllVersions(ident, false, user, false);
+				for (Contentlet con : contentlets) {
+					add(con, PusheableAsset.CONTENTLET);
+				}
+			}
+		}
+
+		try {
+			if (publisherFilter.isDependencies()) {
+				config.waitUntilResolveAllDependencies();
+			}
+		} catch (ExecutionException e) {
+			final Optional<Throwable> causeOptional = ExceptionUtil
+					.getCause(e, set(DotBundleException.class));
+
+			if (causeOptional.isPresent()) {
+				throw (DotBundleException) causeOptional.get();
+			} else {
+				final Throwable rootCause = ExceptionUtil.getRootCause(e);
+				throw new DotBundleException(rootCause.getMessage(), (Exception) rootCause);
+			}
+		}
+	}
+
+	private <T> void add(final T asset, final PusheableAsset pusheableAsset) {
+		final boolean added = config.addWithDependencies(asset, pusheableAsset);
+
+		if (added) {
+			pushedAssetUtil.savePushedAssetForAllEnv(asset, pusheableAsset);
+		}
+	}
+
+	@VisibleForTesting
+	public Set getContents() {
+		return config.getContentlets();
+
+	}
+
+	@VisibleForTesting
+	public Set getRelationships() {
+		return config.getRelationships();
+	}
+
+	@VisibleForTesting
+	public Set getTemplates() {
+		return config.getTemplates();
+	}
+
+	@VisibleForTesting
+	public Set getContainers() {
+		return config.getContainers();
+	}
+
+	@VisibleForTesting
+	public Set getFolders() {
+		return config.getFolders();
+	}
+
+	public static <T> String getKey ( final T asset) {
+		if (Contentlet.class.isInstance(asset)) {
```

Comment:
Not sure if this should be in another method, also not sure if @fabrizzio-dotCMS may have a better idea where to encapsulate this

Reply:
done https://github.com/dotCMS/core/pull/20521/commits/078b38ddfdf91f36ed147df16700c696a0da57c1#diff-b2d9b2ef3027cf0691e8532ead5ebd45917b52beae669c039b06fb14eef45600R258